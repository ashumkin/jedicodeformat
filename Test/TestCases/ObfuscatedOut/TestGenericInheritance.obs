program testgenericinheritahnce;{$APPTYPE CONSOLE}uses sysutils;type ttest=class
fdata:integer;end;ttest<i>=class fdata:i;end;ttest<i,s>=class fdata:i;fitem:s;end
;tgeneric<a,b>=class fdata:a;fitem:b;end;tgeneric<a>=class(TGeneric<A,String>)fdata
:a;end;tgenerictest<r,u>=class(TGeneric<R>)fdata:r;end;var gen1:ttest;gen2:ttest<
integer>;gen3:ttest<integer,string>;begin try gen1:=ttest.create;gen2:=ttest<integer
>.create;gen3:=ttest<integer,string>.create;except on e:exception do writeln(e.classname
,': ',e.message);end;end. 