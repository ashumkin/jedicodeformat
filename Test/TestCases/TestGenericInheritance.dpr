program TestGenericInheritahnce;

{$APPTYPE CONSOLE}

{ AFS December 2007 

  This code compiles, but is not semantically meaningfull.
  It is test cases for the code-formating utility

  Test new generics syntax - code from TridenT
}

uses
  SysUtils;
type

  TTest=Class
   FData: integer;
  End;
  //Surcharge générique
  TTest<I>=Class
   FData: I;
  End;

  {La surcharge basée sur une contrainte est impossible.
  TTest<I:constructor>=Class //E2037 La déclaration de 'TTest<I>' diffère de la déclaration précédente
   FData: I;
  End;
  }

  TTest<I,S>=Class
   FData: I;
   Fitem: S;
  End;

  //Surcharge générique à partir d'une autre surcharge générique de la même classe
      TGeneric<A,B>=Class
       FData: A;
       Fitem: B;
      End;

  TGeneric<A>=Class(TGeneric<A,String>)
   FData: A;
  End;

  //Surcharge générique à partir d'une autre surcharge générique d'une autre classe
  TGenericTest<R,U>=Class(TGeneric<R>)
   FData: R;
  End;

  {La surcharge basée sur une type valeur au lieu d'un type référence est impossible.
   TGenericTest<R,U>=Record //E2004 Identificateur redéclaré : 'TGenericTest'

  End;
  }
var Gen1 : TTest;
    Gen2 : TTest<Integer>;
    Gen3 : TTest<Integer,String>;
begin
  try
   Gen1:= TTest.Create;
   Gen2:= TTest<Integer>.Create;
   Gen3:= TTest<Integer,String>.Create;
  except
    on E:Exception do
      Writeln(E.Classname, ': ', E.Message);
  end;
end.
