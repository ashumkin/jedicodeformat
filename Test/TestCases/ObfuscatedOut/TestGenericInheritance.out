program testgenericinheritahnce;{$APPTYPE CONSOLE}

uses
  SysUtils;

type
  ttest = class
    fdata: integer;
  end;

  ttest<i> = class
    fdata: i;
  end;

  ttest<i, s> = class
    fdata: i;
    fitem: s;
  end;

  tgeneric<a, b> = class
    fdata: a;
    fitem: b;
  end;

  tgeneric<a> = class(TGeneric<A, string>)
    fdata: a;
  end;

  tgenerictest<r, u> = class(TGeneric<R>)
    fdata: r;
  end;
var
  gen1: ttest;
  gen2: ttest<integer>;
  gen3: ttest<integer, string>;
begin
  try
    gen1 := ttest.Create;
    gen2 := ttest<integer>.Create;
    gen3 := ttest<integer, string>.Create;
  except
    on e: Exception do
      writeln(e.ClassName, ': ', e.message);
  end;
end.
 